examples/next/app-router/app/login/page.tsx

```diff
@@ -0,0 +1,47 @@
+import { onLogin } from '@faustwp/experimental-app-router';
+import { redirect } from 'next/navigation';
+
+export default async function Page() {
+  async function loginAction(formData: FormData) {
+    'use server';
+
+    const res = await onLogin(formData);
+
+    if (res.error) {
+      /**
+       * @TODO Next.js is still working on ways persisting error messages from
+       * server actions to the client.
+       *
+       * "Displaying loading or error states currently requires using
+       * Client Components. We are exploring options for server-side functions
+       * to retrieve these values as we move forward in stability for Server Actions."
+       *
+       * @link https://nextjs.org/docs/app/building-your-application/data-fetching/forms-and-mutations#error-handling
+       */
+      console.error(res.error);
+    } else {
+      redirect('/gated-content');
+    }
```

Comment:
This is annoying. Next.js server actions don't yet have a way to persist responses from server actions to the client:

> Good to know:

> Displaying loading or error states currently requires using Client Components. We are exploring options for server-side functions to retrieve these values as we move forward in stability for Server Actions.

Although they say they are working on it.

https://nextjs.org/docs/app/building-your-application/data-fetching/forms-and-mutations#error-handling